File: packages/build-worker/src/services/builder.ts
import { GitService } from './git-service';
import { LogStreamer } from './log-streamer';
import { ArtifactService } from './artifact-service';
import { join } from 'path';
import { spawn } from 'child_process';
import { rm } from 'fs/promises';

interface BuildJob {
  build_id: string;
  project_id: string;
  github_url: string;
  build_command: string;
  root_directory: string;
  app_type: 'nextjs' | 'vite';
  env_vars: Record<string, string>;
}

export const Builder = {
  async execute(job: BuildJob) {
    console.log(`[Builder] Starting execution for build ${job.build_id}`);
    const workDir = join(process.cwd(), 'workspace', job.build_id);
    const controlApiUrl = process.env.CONTROL_API_URL || 'http://localhost:4000';

    const updateStatus = async (status: 'building' | 'success' | 'failed') => {
      try {
        await fetch(`${controlApiUrl}/builds/${job.build_id}`, {
          method: 'PUT',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ status }),
        });
      } catch (error) {
        console.error('Failed to update build status:', error);
      }
    };

    try {
      await updateStatus('building');

      // 1. Clone
      await LogStreamer.stream(
        job.build_id,
        job.project_id,
        `Starting build for ${job.build_id}\n`,
      );

      let token: string | undefined;
      // Resolve installation token if installation_id exists
      if ((job as any).installation_id) {
        // Cast to any because we updated interface in control-api but strict types might miss here unless we update build-worker interface too.
        try {
          const { GitHubService } = await import('./github-service');
          await LogStreamer.stream(
            job.build_id,
            job.project_id,
            'Authenticating with GitHub App...\n',
          );
          token = await GitHubService.getInstallationToken((job as any).installation_id);
        } catch (e: any) {
          await LogStreamer.stream(
            job.build_id,
            job.project_id,
            `GitHub Auth Failed: ${e.message}\n`,
          );
          // Proceed? No build will fail if private.
          throw e;
        }
      }

      await LogStreamer.stream(job.build_id, job.project_id, 'Cloning repository...\n');
      await GitService.clone(job.github_url, workDir, token);

      // 2. Install
      await LogStreamer.stream(job.build_id, job.project_id, 'Installing dependencies...\n');
      const projectDir = join(workDir, job.root_directory);
      await this.runCommand('bun install', projectDir, job.build_id, job.project_id, job.env_vars);

      // 3. Build
      await LogStreamer.stream(job.build_id, job.project_id, 'Building project...\n');
      await this.runCommand(
        job.build_command,
        projectDir,
        job.build_id,
        job.project_id,
        job.env_vars,
      );

      await LogStreamer.stream(job.build_id, job.project_id, 'Build completed successfully!\n');
      await LogStreamer.stream(job.build_id, job.project_id, 'Creating artifact package...\n');
      await LogStreamer.stream(
        job.build_id,
        job.project_id,
        'Streaming artifact to Deploy Engine...\n',
      );

      await ArtifactService.streamArtifact(job.build_id, projectDir, job.app_type);

      await LogStreamer.stream(job.build_id, job.project_id, 'Artifact uploaded successfully!\n');

      await updateStatus('success');
    } catch (error: any) {
      await LogStreamer.stream(job.build_id, job.project_id, `Build failed: ${error.message}\n`);
      await updateStatus('failed');
      throw error;
    } finally {
      await LogStreamer.ensureFlushed(job.build_id);
      try {
        await rm(workDir, { recursive: true, force: true });
        console.log(`Cleaned up workspace: ${workDir}`);
      } catch (e) {
        console.error(`Failed to cleanup workspace: ${workDir}`, e);
      }
    }
  },

  async runCommand(
    command: string,
    cwd: string,
    buildId: string,
    projectId: string,
    envVars: Record<string, string> = {},
  ) {
    return new Promise<void>((resolve, reject) => {
      const [cmd, ...args] = command.split(' ');
      const child = spawn(cmd, args, {
        cwd,
        shell: true,
        env: { ...process.env, ...envVars },
      });

      child.stdout.on('data', (data) => {
        LogStreamer.stream(buildId, projectId, data.toString());
      });

      child.stderr.on('data', (data) => {
        LogStreamer.stream(buildId, projectId, data.toString());
      });

      child.on('close', (code) => {
        if (code === 0) {
          resolve();
        } else {
          reject(new Error(`Command exited with code ${code}`));
        }
      });

      child.on('error', (err) => {
        reject(err);
      });
    });
  },
};

File: packages/build-worker-go/internal/services/builder.go
package services

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"sync"

	"github.com/thakurdotdev/build-worker/internal/config"
	"github.com/thakurdotdev/build-worker/internal/utils"
)

// BuildJob represents a build request
type BuildJob struct {
	BuildID        string            `json:"build_id"`
	ProjectID      string            `json:"project_id"`
	GithubURL      string            `json:"github_url"`
	BuildCommand   string            `json:"build_command"`
	RootDirectory  string            `json:"root_directory"`
	AppType        string            `json:"app_type"`
	EnvVars        map[string]string `json:"env_vars"`
	InstallationID string            `json:"installation_id,omitempty"`
}

// BuilderPool manages concurrent build workers
type BuilderPool struct {
	jobs    chan BuildJob
	workers int
	wg      sync.WaitGroup
}

var builderPool *BuilderPool
var poolOnce sync.Once

// GetBuilderPool returns the singleton pool
func GetBuilderPool() *BuilderPool {
	poolOnce.Do(func() {
		cfg := config.Get()
		builderPool = &BuilderPool{
			jobs:    make(chan BuildJob, 100),
			workers: cfg.BuildWorkers,
		}
		builderPool.start()
	})
	return builderPool
}

// start launches worker goroutines
func (p *BuilderPool) start() {
	for i := 0; i < p.workers; i++ {
		p.wg.Add(1)
		go p.worker(i)
	}
	fmt.Printf("[Builder] Started %d workers\n", p.workers)
}

// worker processes jobs from the channel
func (p *BuilderPool) worker(id int) {
	defer p.wg.Done()
	for job := range p.jobs {
		fmt.Printf("[Worker %d] Processing build %s\n", id, job.BuildID)
		if err := ExecuteBuild(job); err != nil {
			fmt.Printf("[Worker %d] Build %s failed: %v\n", id, job.BuildID, err)
		}
	}
}

// Submit adds a job to the queue
func (p *BuilderPool) Submit(job BuildJob) {
	p.jobs <- job
}

// ExecuteBuild runs the full build pipeline
func ExecuteBuild(job BuildJob) error {
	cfg := config.Get()
	workDir := filepath.Join(cfg.WorkspaceDir, job.BuildID)
	ctx := context.Background()

	log := func(msg string, level LogLevel) {
		StreamLog(job.BuildID, job.ProjectID, msg, level)
	}

	updateStatus := func(status string) {
		url := fmt.Sprintf("%s/builds/%s", cfg.ControlAPIURL, job.BuildID)
		payload, _ := json.Marshal(map[string]string{"status": status})
		req, _ := http.NewRequest("PUT", url, bytes.NewReader(payload))
		req.Header.Set("Content-Type", "application/json")
		http.DefaultClient.Do(req)
	}

	defer func() {
		EnsureFlushed(job.BuildID)
		// Cleanup workspace
		os.RemoveAll(workDir)
		fmt.Printf("[Builder] Cleaned up: %s\n", workDir)
	}()

	updateStatus("building")
	log(fmt.Sprintf("Starting build for %s\n", job.BuildID), LogLevelInfo)

	// 1. Authenticate with GitHub if needed
	var token string
	if job.InstallationID != "" {
		log("Authenticating with GitHub App...\n", LogLevelInfo)
		ghService, err := GetGitHubService()
		if err != nil {
			log(fmt.Sprintf("GitHub Auth Failed: %v\n", err), LogLevelError)
			updateStatus("failed")
			return err
		}
		token, err = ghService.GetInstallationToken(job.InstallationID)
		if err != nil {
			log(fmt.Sprintf("GitHub Auth Failed: %v\n", err), LogLevelError)
			updateStatus("failed")
			return err
		}
	}

	// 2. Clone repository
	log("Cloning repository...\n", LogLevelInfo)
	if err := GetGitService().Clone(ctx, job.GithubURL, workDir, token); err != nil {
		log(fmt.Sprintf("Clone failed: %v\n", err), LogLevelError)
		updateStatus("failed")
		return err
	}

	projectDir := filepath.Join(workDir, job.RootDirectory)

	// 3. Handle build based on framework
	if isBackendFramework(job.AppType) {
		needsBuild := utils.NeedsCompilationStep(job.BuildCommand)
		hasBuildScript := utils.HasScript(projectDir, "build")

		if needsBuild && hasBuildScript {
			log("TypeScript backend detected - running build step...\n", LogLevelInfo)
			log("Installing dependencies...\n", LogLevelInfo)
			result := utils.RunCommand(ctx, "bun install", projectDir, job.EnvVars, func(line string) {
				log(line, LogLevelInfo)
			})
			if result.ExitCode != 0 {
				log(fmt.Sprintf("Install failed with exit code %d\n", result.ExitCode), LogLevelError)
				updateStatus("failed")
				return fmt.Errorf("install failed")
			}

			log("Building project...\n", LogLevelInfo)
			result = utils.RunCommand(ctx, job.BuildCommand, projectDir, job.EnvVars, func(line string) {
				log(line, LogLevelInfo)
			})
			if result.ExitCode != 0 {
				log(fmt.Sprintf("Build failed with exit code %d\n", result.ExitCode), LogLevelError)
				updateStatus("failed")
				return fmt.Errorf("build failed")
			}
			log("Build completed successfully!\n", LogLevelSuccess)
		} else {
			log("Backend project detected - skipping build step...\n", LogLevelInfo)
			log("Source code will be packaged and dependencies installed at deploy time.\n", LogLevelInfo)
		}
	} else {
		// Frontend: always install and build
		log("Installing dependencies...\n", LogLevelInfo)
		result := utils.RunCommand(ctx, "bun install", projectDir, job.EnvVars, func(line string) {
			log(line, LogLevelInfo)
		})
		if result.ExitCode != 0 {
			log(fmt.Sprintf("Install failed with exit code %d\n", result.ExitCode), LogLevelError)
			updateStatus("failed")
			return fmt.Errorf("install failed")
		}

		log("Building project...\n", LogLevelInfo)
		result = utils.RunCommand(ctx, job.BuildCommand, projectDir, job.EnvVars, func(line string) {
			log(line, LogLevelInfo)
		})
		if result.ExitCode != 0 {
			log(fmt.Sprintf("Build failed with exit code %d\n", result.ExitCode), LogLevelError)
			updateStatus("failed")
			return fmt.Errorf("build failed")
		}
		log("Build completed successfully!\n", LogLevelSuccess)
	}

	// 4. Create and upload artifact
	log("Creating artifact package...\n", LogLevelInfo)
	log("Streaming artifact to Deploy Engine...\n", LogLevelInfo)

	if err := GetArtifactService().StreamArtifact(job.BuildID, projectDir, job.AppType); err != nil {
		log(fmt.Sprintf("Artifact upload failed: %v\n", err), LogLevelError)
		updateStatus("failed")
		return err
	}

	log("Artifact uploaded successfully!\n", LogLevelSuccess)
	updateStatus("success")
	return nil
}
